Fixes for nemoLite2d OpenCL generation (for #495, #494, ...)

sergisiso opened the pull request:
A few small fixes we need to have nemoLite2d working with the OpenCL back-end:
- [x] Fix C type casting operation #494 
- [x] ~~Kernel renaming with kernel name #499~~ (temporal Quickfix for OpenCL kernels)
- [x] Add local-size and command-queue parameters in each kernel #495
- [x] Remove duplication of width parameter and global_work_size()

sergisiso commented:
@arporter @rupertford This is ready for its first review. It contains the necessary changes to make the OpenCL generation work automatically (with some caveats which are marked with TODOs) for the nemolite2d application (it has been tested with the PSycloneBench opencl_psykal branch and it produces the same results than the CPU serial version). It also provides a set of options to start exploring OpenCL optimizations.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -1768,15 +1800,21 @@ def gen_code(self, parent):
+
+            # We need a clFinish for all the queues in the implementation
+            opencl_num_queues = 1
+            for kern in self.coded_kernels():
```
This bit of code working out the number of queues is replicated from earlier. Could it be a method (or have the result stored somewhere)?

sergisiso replied:
They are not exactly the same as the one for the opencl_init gets the maximum of all Invokes to allocate the resources, while this one is the maximum for the specific Invoke because it just needs barriers for the queues it uses.

Is it true that the former could include a call to the latter, but it will still be two max() operations very similar to the current implementations, so I think it's better to keep it this way for now.

I also prefer not to have this value stored as even if it is not implemented yet I like the idea of some 'smart' scheduling transformation that changes the queues of each kernel at its convenience and it is only when we call `gen_code` that this has to be fully determined.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -1768,15 +1800,21 @@ def gen_code(self, parent):
+                opencl_num_queues = max(
+                    opencl_num_queues,
+                    kern._opencl_options['queue_number'])
+            for i in range(1, opencl_num_queues + 1):
```
I suspect pylint/pep8 won't like `i`.

sergisiso replied:
Of course, updated to `queue_number` :)

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -3773,6 +3812,38 @@ def get_kernel_schedule(self):
 
+    def set_opencl_options(self, options):
+        '''
+        Validate and store a set of options associated to the Kernel to
```
s/associated to/associated with/

sergisiso replied:
Fixed

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -3773,6 +3812,38 @@ def get_kernel_schedule(self):
+        tune the OpenCL code generation.
+
+        :param options: a set of options to tune the OpenCL code.
+        :type options: map of <string>:<value>
```
s/map/dict/

sergisiso replied:
Fixed

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -3773,6 +3812,38 @@ def get_kernel_schedule(self):
+            else:
+                raise AttributeError(
+                    "CodedKern does not support the opencl_option '{0}'."
+                    "".format(key))
```
Please include the valid options in this error message.

sergisiso replied:
Done

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -3773,6 +3812,38 @@ def get_kernel_schedule(self):
+
+            self._opencl_options[key] = value
+
+        # If options are valid copy them in the associated KernelSchedule
```
Picky but s/in the/into the/
Also, why do we need a copy of these properties in the KernelSchedule? Is it possible to just have one single place to keep them?

sergisiso replied:
Removed these 3 lines to avoid duplicating the Kernel options.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -3977,7 +4048,23 @@ def rename_and_write(self):
-            new_suffix = "_{0}".format(name_idx)
+            new_suffix = ""
+
+            # On the GOcean OpenCL work we need to differentiate between the
```
"...between kernels generated from the same module file..."

sergisiso replied:
Done

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -5963,6 +6052,17 @@ def symbol_table(self):
+    def set_opencl_options(self, options):
+        '''
+        Set the opencl_options map. Note that KernelSchedule expects to
+        receive options which are already vaildated by the caller.
```
I can see where you're coming from but I think it's better to be defensive and do the checks here too. Having said that, it would be even better if we only kept these properties in one place.

sergisiso replied:
Removed this method as agreed in the conversation below.

arporter commented on src/psyclone/psyir/backend/opencl.py:
```diff
@@ -162,9 +162,18 @@ def kernelschedule_node(self, node):
+        if node._opencl_options['local_size'] != 1:
+            loc_size_value = node._opencl_options['local_size']
+            code += "__attribute__((vec_type_hint(double)))\n" \
+                    "".format(loc_size_value)
```
You're using `.format()` here but there's no e.g. `{}` or `{0}` in the string itself.

sergisiso replied:
Ups, fixed.

arporter commented on src/psyclone/tests/gocean1p0_opencl_test.py:
```diff
@@ -327,6 +461,7 @@ def test_symtab_implementation_for_opencl():
         in str(err)
 
 
+@pytest.mark.xfail(reason="Use statment check is bypassed to make OpenCL work")
```
Please label this with an Issue number.

sergisiso replied:
Done

arporter commented on src/psyclone/transformations.py:
```diff
@@ -2285,46 +2285,57 @@ def apply(self, sched, opencl=True):
         sched.opencl = opencl
+
+        # This transformation stores some options into the InvokeSchedule
+        sched.set_opencl_options(options)
```
Since this call can raise Exceptions I think we should catch them here and re-raise them as a `TransformationError` for consistency with other transformations.

sergisiso replied:
Done

sergisiso commented on doc/user_guide/transformations.rst:
```diff
@@ -528,15 +528,76 @@ of the FortCL library (https://github.com/stfc/FortCL) to access
+
+The ``OCLTrans`` transformation accepts an `options` argument with a
+map of optional parameters to tune the OpenCL host code in the PSy layer.
+This options will be attached to the transformed InvokeSchedule.
```
Fixed

sergisiso commented on doc/user_guide/transformations.rst:
```diff
@@ -528,15 +528,76 @@ of the FortCL library (https://github.com/stfc/FortCL) to access
+|              | Invoke.                                      |         |
++--------------+----------------------------------------------+---------+
+
+Additionally, each individual kernel inside the Invoke that is going
```
You meant delete the part inside the parenthesis right? In this case Done.

arporter replied:
No, sorry, that's not what I meant. I meant put the text "inside the Invoke that is going to be transformed" inside parentheses :-)

sergisiso replied:
I got you

sergisiso commented on doc/user_guide/transformations.rst:
```diff
@@ -528,15 +528,76 @@ of the FortCL library (https://github.com/stfc/FortCL) to access
+Additionally, each individual kernel inside the Invoke that is going
+to be transformed, also accepts a map of options which are provided by
+the `set_opencl_options()` method of the `Kern` object.
+This can affect both, the driver layer or the OpenCL kernels.
```
Fixed

sergisiso commented on doc/user_guide/transformations.rst:
```diff
@@ -528,15 +528,76 @@ of the FortCL library (https://github.com/stfc/FortCL) to access
+| local_size   | Number of work-items to compute             | 1       |
+|              | in a single kernel.                         |         |
++--------------+---------------------------------------------+---------+
+| queue_number | Queue identifier number where               | 1       | 
```
Replaced text

sergisiso commented on doc/user_guide/transformations.rst:
```diff
@@ -528,15 +528,76 @@ of the FortCL library (https://github.com/stfc/FortCL) to access
++--------------+---------------------------------------------+---------+
+
+
+See below an example of an PSyclone script that uses an ``OCLTans`` with
```
Done

sergisiso commented on doc/user_guide/transformations.rst:
```diff
@@ -528,15 +528,76 @@ of the FortCL library (https://github.com/stfc/FortCL) to access
+multiple InvokeSchedule and kernel-specific transformations.
+
+
+.. code-block:: python
```
Yes, the `.. literalinclude:: ../../examples/gocean/eg3/ocl_trans.py` command did the trick

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -429,23 +429,28 @@ def gen_code(self, parent):
+                generate_ocl_init = True
                 for kern in invoke.schedule.coded_kernels():
                     if kern.name not in opencl_kernels:
+                        opencl_num_queues = max(
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -1661,8 +1668,33 @@ def __init__(self, KernFactory, BuiltInFactory, alg_calls=None):
 
+    def set_opencl_options(self, options):
+        '''
+        Validate and store a set of options associated to the InvokeSchedule to
```
fixed

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -1661,8 +1668,33 @@ def __init__(self, KernFactory, BuiltInFactory, alg_calls=None):
+        tune the OpenCL code generation.
+
+        :param options: a set of options to tune the OpenCL code.
+        :type options: map of <string>:<value>
```
updated docstring

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -1661,8 +1668,33 @@ def __init__(self, KernFactory, BuiltInFactory, alg_calls=None):
+                        "should be a boolean.")
+            else:
+                raise AttributeError(
+                    "InvokeSchedule does not support the opencl_option '{0}'."
```
Done. I used the plural tense "Accepted options are: 'end_barrier'." as this may accept more options if the future. Is this right? or should I use the singular form for now?

sergisiso commented:
@arporter I do agree that the replicated Kern/KernelSchedule opencl_options is not a good design. Looking into it I think they should be in Kern call as not all options are needed in the KSchedule (e.g. num_queue). And in the future we may have multiple calls referencing the same kernel and with distinct opencl_options.

The problem is KernelSchedule does not have a reference to its callee to query the information. In reality we just need to have the 'local_size' (blocking parameter) and there is a few solutions for this: 

1. Add a reference from KernelSchedule to its callee. This is possible because right now each call parses its own instance of the kernel so there is a 1-to-1 match. But I don't like this option as I think each KernelSchedule should be independent of Kern and have the possibility to reuse them.
2.  Add just 'opencl_local_size' as an annotation to KernelSchedule. We would reuse the annotations infrastructure and it fits well with the end result we want which is an '__attribute_(...)' which is kind of a C annotation.
3. When we call the OpenCL visitor from Kern, we can specify an argument `ocl_writer = OpenCLWriter(kernels_local_size=4); ocl_writer(self.get_kernel_schedule())`. This works because we instantiate an OpenCLWriter for each kernel we call.

I am inclined to option 2, but I would like to know your opinion before implementing it.

rupertford commented:
Would it make sense for a KernelSchedule to have a reference to 1 or more callees i.e. implement option 1 but with a 1..n relationship?

arporter commented:
I think I'm a bit confused about the relationship between a `Kern` and a `KernelSchedule`. A given `Kern` must have only one `KernelSchedule` but multiple `Kern`s could all refer to the same `KernelSchedule`. However, in the latter case, we could conceivably want different local sizes for different kernel calls? Since that has to be hard-coded at compile time we would then need to generate different versions of the same kernel. That then feels like the kernels are in fact different and this should be reflected by having different `Kern` objects in the Schedule. Is that correct?

rupertford commented:
I think the idea of multiple Kerns referencing the same KernelSchedule would be a new concept in PSyclone. I think that at the moment we assume a 1-1 relationship internally (but we do something when writing them out to file with the 'single' etc options).

If we were to go down this route then all changes to KernelSchedule would be "seen" by all Kerns that reference it and if we wanted different things to happen in different contexts we would have to have a transformation to split/replicate the KernelSchedule (giving it a different name/instance id).

sergisiso commented:
That's right, at the moment each Kern generate/parses its associated KernelSchedule and maintains a reference Kern->KernelSchedule, so Kerns pointing to the same PSyKAl kernels will have individual KernelSchedules.

I am not proposing changing this implementation but I want to avoid making a Kern<-KernelSchedule reference as I think it does not fit conceptually even if the implementation would work.

This is why I think options 2 and 3 are better (although they have their own problems).

arporter commented:
Thanks for clarifying. I'm not keen on option 2 as it feels like a re-purposing of `annotations`: currently we only use them to store information on where Nodes in the AST came from whereas with this change they would affect the generated code. I quite like option 3 - what are the drawbacks you see with that approach?

sergisiso commented:
After thinking again about this I now also prefer option 3, so it is the one that I will implement.

The good things about 3 is that it is specific of the OpenCL backend, so the data is just where it is needed and as currently implemented the Writer instacnce is short-lived (just created for generating 1 kernel and then destroyed, so the duplication of the opencl options are not permanent).

At first I though it was odd to give a parameter that just the kernel node cares about to the general back-end. But now I think this is because openCL conveniently has the special purpose attribute but if not it could affect multiple nodes.

sergisiso commented:
@arporter This is ready for the next review.

The main change is the removal of the KernelSchedule opencl_options duplication and I added instead a OpenCLWriter argument to provide the 'local_size'. I addressed all other inline comments but the one requesting a method to compute the maximum number of queues, see justification in-line.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -1768,15 +1803,22 @@ def gen_code(self, parent):
+            for kern in self.coded_kernels():
+                opencl_num_queues = max(
+                    opencl_num_queues,
+                    kern._opencl_options['queue_number'])
```
pylint complains about this reference to `_opencl_options` - I guess we need a getter.

sergisiso replied:
Added a property to access opencl_options

arporter commented on src/psyclone/tests/gocean1p0_opencl_test.py:
```diff
@@ -254,13 +359,43 @@ def test_opencl_kernel_code_generation():
+    otrans = OCLTrans()
+    otrans.apply(sched)
+    sched.kernels()[0].name = "name"
+    generated_code = str(psy.gen).lower()
```
pylint complains that `generated_code` is unused. We could just remove the assignment part of this statement altogether (as I think we've agreed/voted to not use `_` in tests unless absolutely necessary).

sergisiso replied:
This is needed because it generates the OpenCL kernel as side-effect. I removed the `str().lower()`, used `_` and added a comment mentioning the side-effect.

arporter replied:
Great, thanks.

arporter commented on src/psyclone/tests/gocean1p0_opencl_test.py:
```diff
@@ -254,13 +359,43 @@ def test_opencl_kernel_code_generation():
+    sched = psy.invokes.invoke_list[0].schedule
+    otrans = OCLTrans()
+    otrans.apply(sched)
+    generated_code = str(psy.gen).lower()
```
Again, `generated_code` is unused.

sergisiso replied:
Same as above

arporter commented on src/psyclone/tests/gocean1p0_transformations_test.py:
```diff
@@ -1475,7 +1475,8 @@ def test_ocl_apply(kernel_outputdir):
         _, _ = ocl.apply(schedule.children[0])
-    assert "the supplied node must be a (sub-class of) Schedule " in str(err)
+    assert "the supplied node must be a (sub-class of) InvokeSchedule " \
+        "" in str(err)
```
I don't think you need the '""' on this line.

sergisiso replied:
Ups, deleted.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -2265,7 +2265,7 @@ def name(self):
         return "OCLTrans"
 
-    def apply(self, sched, opencl=True):
+    def apply(self, sched, opencl=True, options={}):
```
I should have picked this up first time around but pylint warns that `{}` is a dangerous default value. Please use None instead and change the method body appropriately.

sergisiso replied:
Fixed

sergisiso commented on doc/user_guide/transformations.rst:
```diff
@@ -561,33 +561,19 @@ The current available options are:
 
-See below an example of an PSyclone script that uses an ``OCLTans`` with
-multiple InvokeSchedule and kernel-specific transformations.
+Below is an example of an PSyclone script that uses an ``OCLTans`` with
```
Fixed

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -1690,7 +1692,8 @@ def set_opencl_options(self, options):
                 raise AttributeError(
-                    "InvokeSchedule does not support the opencl_option '{0}'."
+                    "InvokeSchedule does not support the opencl_option '{0}'. "
+                    "The supported options are: 'end_barrier'."
```
Done for the 1696, but I am not sure if I should do it for 1688 as the error is specific to that key.

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -3835,15 +3839,12 @@ def set_opencl_options(self, options):
                 raise AttributeError(
-                    "CodedKern does not support the opencl_option '{0}'."
-                    "".format(key))
+                    "CodedKern does not support the opencl_option '{0}'. "
```
Done

sergisiso commented on src/psyclone/psyir/backend/opencl.py:
```diff
@@ -47,7 +47,38 @@ class OpenCLWriter(CWriter):
+                "'{0}'.".format(type(kernels_local_size).__name__))
+
+        if kernels_local_size < 1:
+            raise TypeError(
```
Done

sergisiso commented on src/psyclone/psyir/backend/opencl.py:
```diff
@@ -163,12 +194,10 @@ def kernelschedule_node(self, node):
-            code += "__attribute__((vec_type_hint(double)))\n" \
-                    "".format(loc_size_value)
+        if self._kernels_local_size != 1:
+            code += "__attribute__((vec_type_hint(double)))\n"
```
I will delete this line for now, although recommended in the Xilinx User Guide I didn't observe any performance difference.

I explore this further and re-introduce it in the future if necessary.

sergisiso commented:
@arporter If the unit-test passes, this is ready for next review. I also check again that the OpenCL still produces exact same results as the serial version for NemoLite2D after the changes.

arporter commented:
All requested changes have been made. All tests still pass/xfail, including with compilation.

arporter commented:
Once #511 is merged, this is ready to go (needs changelog and UG update).
